compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -641,7 +659,37 @@ pub fn orphan_check(tcx: TyCtxt<'_>, impl_def_id: DefId) -> Result<(), OrphanChe
         return Ok(());
     }
 
-    orphan_check_trait_ref::<!>(trait_ref, InCrate::Local, |ty| Ok(ty)).unwrap()
+    orphan_check_trait_ref::<!>(trait_ref, InCrate::Local, |raw_ty| {
+        let ty::Alias(..) = raw_ty.kind() else { return Ok((raw_ty, None)) };
+
+        let infcx = tcx.infer_ctxt().intercrate(true).build();
```

Comment:
instead of creating a new inference context here, create the context outside of this function and eagerly instantiate the impl with fresh args :thinking: